tests/validation/test_cat_vrs_models.py

```diff
@@ -0,0 +1,350 @@
+        recipes.CanonicalAllele(**valid_params)
+
+    # Invalid CanonicalAllele: > 1 'transcribes_to'
+    valid_params = deepcopy(members)
+    valid_params["constraints"] = [
+        models.Constraint(
+            root=models.DefiningAlleleConstraint(
+                relations=[
+                    MappableConcept(
+                        primaryCode=code(models.Relation.LIFTOVER_TO.value)
+                    ),
+                    MappableConcept(
+                        primaryCode=code(models.Relation.TRANSCRIBES_TO.value)
+                    ),
+                    MappableConcept(
+                        primaryCode=code(models.Relation.TRANSCRIBES_TO.value)
+                    ),
+                ],
+                allele="allele.json#/1",
+            )
+        )
+    ]
+    with pytest.raises(
+        ValueError,
+        match="Must contain exactly one relation where `primaryCode` is 'transcribes_to'.",
+    ):
+        recipes.CanonicalAllele(**valid_params)
+
+
+def test_categorical_cnv(members, defining_loc_constr, copy_change_constr):
+    """Test the CategoricalCnv validator"""
+    # Valid CategoricalCnv with CopyChangeConstraint
+    valid_params = deepcopy(members)
+    valid_params["constraints"] = [
+        models.Constraint(root=defining_loc_constr),
+        models.Constraint(root=copy_change_constr),
+    ]
+    assert recipes.CategoricalCnv(**valid_params)
+
+    # Valid CategoricalCnv with CopyCountConstraint
+    valid_params = deepcopy(members)
+    valid_params["constraints"] = [
+        models.Constraint(root=defining_loc_constr),
+        models.Constraint(root=models.CopyCountConstraint(copies=[1, 2])),
+    ]
+    assert recipes.CategoricalCnv(**valid_params)
+
+    # Invalid CategoricalCnv: No DefiningLocationConstraint
+    invalid_params = deepcopy(members)
+    invalid_params["constraints"] = [
+        models.Constraint(root=def_allele_constr_empty_relations()),
+        copy_change_constr,
+    ]
+    with pytest.raises(
+        ValueError,
+        match="Must contain a `DefiningLocationConstraint` with at least one relation where `primaryCode` is 'liftover_to'.",
+    ):
+        recipes.CategoricalCnv(**invalid_params)
+
+    # Invalid CategoricalCnv: DefiningLocationConstraint does not have 'liftover_to'
+    invalid_params = deepcopy(members)
+    invalid_defining_loc_constr = defining_loc_constr.copy(deep=True)
+    invalid_defining_loc_constr.relations = [
+        MappableConcept(primaryCode=code(models.Relation.TRANSCRIBES_TO.value))
+    ]
+    invalid_params["constraints"] = [
+        invalid_defining_loc_constr,
+        copy_change_constr,
+    ]
+    with pytest.raises(
+        ValueError,
+        match="Must contain a `DefiningLocationConstraint` with at least one relation where `primaryCode` is 'liftover_to'.",
+    ):
+        recipes.CategoricalCnv(**invalid_params)
+
+    # Invalid No CopyCountConstraint or CopyChangeConstraint
+    invalid_params["constraints"] = [
```

Comment:
This should instead be

```python
invalid_params["constraints"] = [
    copy_change_constr
]
```

right? This may also require some additional tests to make sure that either `CopyCountConstraint` or `CopyChangeConstraint`, whichever is passed, is valid.

Reply:
I can update the comment to be clearer. This was testing that constraints does not contain either a `CopyCountConstraint` or a `CopyChangeConstraint`. 

The first two examples in `test_categorical_cnv` test valid input (either `CopyCountConstraint` or `CopyChangeConstraint`). I have tests for `CopyChangeConstrint` in `test_copy_change_constraint`, but can make tests for `CopyCountConstraint`.

Reply:
re: comment Ah!! Okay, I see now. Thank you! 

re: tests. that was so quick! thank you!